### PR TITLE
Add resumable uploads support to GCS repository integration tests

### PR DESCRIPTION
**Pull request by tlrx**
This pull request adds support for [resumable uploads](https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload) to the internal HTTP server used in `GoogleCloudStorageBlobStoreRepositoryTests`. This way we can also test the behavior of the Google's client when the service returns server errors in response to resumable upload requests. 

The `BlobStore` implementation for GCS has the choice between 2 methods to upload a blob: `resumable` and `multipart`. In the current implementation, the client executes a resumable upload if the blob size is larger than `LARGE_BLOB_THRESHOLD_BYTE_SIZE`, otherwise it executes a multipart upload. This pull request makes this logic overridable in tests, allowing to randomize the decision of using one method or the other.

The pull request add support for single request resumable uploads and chunked resumable uploads (the blob is uploaded into multiple 2Mb chunks; each chunk being a resumable upload). For this last case, this PR also adds a test `testSnapshotWithLargeSegmentFiles` which makes it more probable that a chunked resumable upload is executed.

### CODE REVIEW

File: plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStoreRepositoryTests.java [hunk 1]
```diff
@@ -313,66 +388,138 @@ public void handle(final HttpExchange exchange) throws IOException {
...
-                                        int next = in.read();
-                                        if (next != -1) {
-                                            if (next == '\n') {
-                                                break;
+                } else if (Regex.simpleMatch("POST /upload/storage/v1/b/bucket/*uploadType=*", request)) {
+                    final Map<String, String> params = new HashMap<>();
+                    RestUtils.decodeQueryString(exchange.getRequestURI().getQuery(), 0, params);
+                    final String uploadType = params.get("uploadType");
+
+                    if (uploadType.equals("resumable")) {
```
Comment (tlrx) [inline plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStoreRepositoryTests.java]: I could maybe avoid this extra condition and make the code easier to read by using two different regex:

```
} else if (Regex.simpleMatch("POST /upload/storage/v1/b/bucket/*uploadType=resumable*", request)) {
...
} else if (Regex.simpleMatch("POST /upload/storage/v1/b/bucket/*uploadType=multipart*", request)) {
...
```

**Review by original-brownbear — COMMENTED** [review]
Just a few questions. This is great in general, big thanks once again for working through all the tricky details here!

**Review by original-brownbear — APPROVED** [review]
LGTM (assuming CI likes it too), thanks Tanguy!

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by tlrx** [comment]
@elasticmachine update branch

**Comment by tlrx** [comment]
Thanks @original-brownbear !

**Pull request merged**